python/phase-2.md

```diff
@@ -0,0 +1,996 @@
+  cancellation scope
+* Do we need extra capabilities such as:
+  * Ability to explicitly add cancellables?
+  * Ability to unset the current instead of just forcing a detached one each time you want to opt out?
+* Should we disallow the `parent` to even be set in the constructor?
+  * Thinking we probably shouldn't allow you to set that
+* Cancellation of a workflow or a scope causes all things started with that to be issued a cancel
+  * Originally we weren't gonna use cancellation scopes but instead reuse the `asyncio` cancel feature but that throws
+    inside of awaited things and we want to cancel whether awaited or not
+  * This means `asyncio` cancellation (and `shield` etc) affects only in-workflow tasks awaited on, _not_ the actual
+    underlying activity/child being executed
+    * We could disable these features of `asyncio` so people aren't confused by their concept of cancellation and ours,
+      but there is still value in cancelling just the tasks
+
+## Runtime
+
+### Workflow Stepped Execution
+
+Notes:
+
+* We are gonna use `asyncio`, but whether we have to use our own event loop remains to be seen
+  * Have shown at https://github.com/cretz/python-determinism-runner basically using the approach from
+    https://stackoverflow.com/a/29797709 that we can run a single step and enforce that we have yielded only at
+    workflow-acceptable spots
+* Do want to prevent use of `asyncio.sleep`, `asyncio.wait_for`, and similar that are based on system time?
+  * This is much clearer than replacing their implementations so that people aren't confused about why a 50ms wait does
+    not work
+  * We may not be able to block everything until we use a sandbox (see next section)
+  * Making alternatives is a tad ugly, e.g. one can use `asyncio.gather()` without timeout but they can't with timeout,
+    instead they'd have to have a sleep or something
+
+### Workflow Deterministic Execution
+
+Notes:
+
+* A full workflow execution sandbox may not occur until phase 3
+  * Some work on doing this via `eval` are at https://github.com/cretz/python-determinism-runner
+* We will, however, go ahead and abstract the "runner" so that we can make it pluggable/opt-out if the caller chooses
+
+## Additional API Concerns
+
+### Activities as Methods
+
+It is very normal to want to have activities as methods instead of top-level functions. For example, say you have:
+
+```python
+class ActivityClass:
+    def __init__(self, some_init_arg: int) -> None:
+        pass
+    
+    @activity.defn
+    def some_activity(self, arg1: str) -> List[str]:
+        return ["Yay"]
+```
+
+You cannot just `workflow.execute_activity(ActivityClass.some_activity, "foo")` because technically the first parameter
+is `self` there. But sometimes they may want to call
+`workflow.execute_activity(ActivityClass.some_activity, ActivityClass("init-arg1"), "foo")` and who are we to say they
```

Comment:
This specifically we really do want to prevent I believe. The workflow calling the activity cannot make assertions about the state of the activity class since it may well be run on another worker. So, allowing them to pass in instance in the workflow code is pretty misleading.

Reply:
In this case it's just passing as a parameter, just as if it was a top-level function and `ActivityClass` was the first parameter instead of "self". At least at compile time, we can't tell them they can't pass the first legitimate parameter of a two-parameter function.